File: Applied Course/4.Problem Solving/3.Problems on Linked List/25.Middle of the Linked List.cpp
#include<bits/stdc++.h>
using namespace std;
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
// Logic-1
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        if(head==NULL || head->next==NULL)
            return head;
        int count=0;
        ListNode *temp=head;
        while(temp)
        {
            count++;
            temp=temp->next;
        }
        int n=count/2;
        temp=head;
        while(n--)
            temp=temp->next;
        return temp;
    }
};
// Logic-2
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode *fast=head,*slow=head;
        while(fast && fast->next)
        {
            fast=fast->next->next;
            slow=slow->next;
        }
        return slow;
    }
};
int main()
{
    // https://leetcode.com/problems/middle-of-the-linked-list/
    return 0;
}

File: Applied Course/4.Problem Solving/3.Problems on Linked List/25.Middle of the Linked List.py
# https://leetcode.com/problems/middle-of-the-linked-list/
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Logic-1
class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        length=0
        temp=head
        while temp!=None:
            length+=1
            temp=temp.next
        length//=2
        temp=head
        while length:
            length-=1
            temp=temp.next
        return temp
# Logic-2
class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        fast=slow=head
        while fast and fast.next:
            fast=fast.next.next
            slow=slow.next
        return slow